Improve the nbgrader pipeline

jhamrick opened the pull request:
This fixes #144! However, it also depends on ipython/ipython#8073, so it can't be merged until that one is merged.

There's a lot of stuff here, but in general, here's how it works:

1. The base class `BaseNbGraderApp` has variables for `student_id`, `assignment_id`, and `notebook_id`. These are used along with the format string `directory_structure` in order to find the notebooks that need converting.
2. There are also variables `nbgrader_step_input` and `nbgrader_step_output`. These must be overridden by subclasses and correspond to the directories for particular steps of the process (e.g. "submitted" or "autograded").
3. I changed the `BaseNbConvertApp` to override several of the methods from `NbConvertApp`, including initializing the notebooks, initializing the resources for a particular notebook, and actually writing that notebook.
4. `nbgrader assign` and `nbgrader autograde` now depend on the database by default.
5. I also added options to `nbgrader assign` and `nbgrader autograde` to add the assignment and student, respectively, if they don't exist.

The one thing I don't quite know how to do is that I'd like some of these config options to be in a file like `nbgrader_config.py` that gets automatically included if it's in the current directory and applies to all of the apps. This works if the config file matches the name of the app (e.g. `nbgrader_assign_config.py`) but apparently not for `nbgrader_config.py`.

Things that still need to be done:
- [x] fix the tests
- [x] update examples in the different apps
- [ ] flesh out documentation

Even though there's still a bit of work to do I wanted to go ahead and make a PR since I have about 90% of this done.

ping @ellisonbg @jdfreder @Carreau

ellisonbg commented:
Awesome, I will have a look soon!

On Tue, Mar 17, 2015 at 11:31 PM, Jessica B. Hamrick <
notifications@github.com> wrote:

> This fixes #144 <https://github.com/jupyter/nbgrader/issues/144>!
> However, it also depends on ipython/ipython#8073
> <https://github.com/ipython/ipython/pull/8073>, so it can't be merged
> until that one is merged.
>
> There's a lot of stuff here, but in general, here's how it works:
>
>    1. The base class BaseNbGraderApp has variables for student_id,
>    assignment_id, and notebook_id. These are used along with the format
>    string directory_structure in order to find the notebooks that need
>    converting.
>    2. There are also variables nbgrader_step_input and
>    nbgrader_step_output. These must be overridden by subclasses and
>    correspond to the directories for particular steps of the process (e.g.
>    "submitted" or "autograded").
>    3. I changed the BaseNbConvertApp to override several of the methods
>    from NbConvertApp, including initializing the notebooks, initializing
>    the resources for a particular notebook, and actually writing that notebook.
>    4. nbgrader assign and nbgrader autograde now depend on the database
>    by default.
>    5. I also added options to nbgrader assign and nbgrader autograde to
>    add the assignment and student, respectively, if they don't exist.
>
> The one thing I don't quite know how to do is that I'd like some of these
> config options to be in a file like nbgrader_config.py that gets
> automatically included if it's in the current directory and applies to all
> of the apps. This works if the config file matches the name of the app
> (e.g. nbgrader_assign_config.py) but apparently not for nbgrader_config.py
> .
>
> I also have a bit more work to do on fixing the tests, but wanted to go
> ahead and make a PR since I have about 90% of this done.
>
> ping @ellisonbg <https://github.com/ellisonbg> @jdfreder
> <https://github.com/jdfreder> @Carreau <https://github.com/Carreau>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/jupyter/nbgrader/pull/154
> Commit Summary
>
>    - Beginning to get the pipeline to work
>    - Rearrange docs
>    - Make it the default to use the db
>    - Update docs
>    - Allow assignment to be passed in as a command line argument
>    - Fix submit and validate apps
>    - Add update_or_create_assignment api method
>    - Add update_or_create_student api method
>    - Add flags to add assignments/students on the fly
>    - Copy all files that are in the assignment directory
>    - Update tests for nbgrader assign
>
> File Changes
>
>    - *M* .gitignore
>    <https://github.com/jupyter/nbgrader/pull/154/files#diff-0> (8)
>    - *M* docs/user_guide/3 - Releasing assignments.ipynb
>    <https://github.com/jupyter/nbgrader/pull/154/files#diff-1> (99)
>    - *M* docs/user_guide/4 - Autograding.ipynb
>    <https://github.com/jupyter/nbgrader/pull/154/files#diff-2> (98)
>    - *M* docs/user_guide/5 - Manual grading.ipynb
>    <https://github.com/jupyter/nbgrader/pull/154/files#diff-3> (42)
>    - *M* docs/user_guide/6 - Returning feedback.ipynb
>    <https://github.com/jupyter/nbgrader/pull/154/files#diff-4> (47)
>    - *R* docs/user_guide/example/source/Problem Set 1/Problem 1.ipynb
>    <https://github.com/jupyter/nbgrader/pull/154/files#diff-5> (0)
>    - *R* docs/user_guide/example/source/Problem Set 1/Problem 2.ipynb
>    <https://github.com/jupyter/nbgrader/pull/154/files#diff-6> (0)
>    - *R* docs/user_guide/example/source/header.ipynb
>    <https://github.com/jupyter/nbgrader/pull/154/files#diff-7> (0)
>    - *R* docs/user_guide/example/submitted/Bitdiddle/Problem Set
>    1/Problem 1.ipynb
>    <https://github.com/jupyter/nbgrader/pull/154/files#diff-8> (0)
>    - *R* docs/user_guide/example/submitted/Bitdiddle/Problem Set
>    1/Problem 2.ipynb
>    <https://github.com/jupyter/nbgrader/pull/154/files#diff-9> (0)
>    - *R* docs/user_guide/example/submitted/Hacker/Problem Set 1/Problem
>    1.ipynb <https://github.com/jupyter/nbgrader/pull/154/files#diff-10>
>    (0)
>    - *R* docs/user_guide/example/submitted/Hacker/Problem Set 1/Problem
>    2.ipynb <https://github.com/jupyter/nbgrader/pull/154/files#diff-11>
>    (0)
>    - *D* docs/user_guide/grade_example/nbgrader_formgrade_config.py
>    <https://github.com/jupyter/nbgrader/pull/154/files#diff-12> (6)
>    - *M* nbgrader/__init__.py
>    <https://github.com/jupyter/nbgrader/pull/154/files#diff-13> (7)
>    - *M* nbgrader/api.py
>    <https://github.com/jupyter/nbgrader/pull/154/files#diff-14> (62)
>    - *M* nbgrader/apps/assignapp.py
>    <https://github.com/jupyter/nbgrader/pull/154/files#diff-15> (19)
>    - *M* nbgrader/apps/autogradeapp.py
>    <https://github.com/jupyter/nbgrader/pull/154/files#diff-16> (24)
>    - *M* nbgrader/apps/baseapp.py
>    <https://github.com/jupyter/nbgrader/pull/154/files#diff-17> (197)
>    - *M* nbgrader/apps/feedbackapp.py
>    <https://github.com/jupyter/nbgrader/pull/154/files#diff-18> (20)
>    - *M* nbgrader/apps/formgradeapp.py
>    <https://github.com/jupyter/nbgrader/pull/154/files#diff-19> (41)
>    - *M* nbgrader/apps/submitapp.py
>    <https://github.com/jupyter/nbgrader/pull/154/files#diff-20> (8)
>    - *M* nbgrader/apps/validateapp.py
>    <https://github.com/jupyter/nbgrader/pull/154/files#diff-21> (29)
>    - *D* nbgrader/exporters/__init__.py
>    <https://github.com/jupyter/nbgrader/pull/154/files#diff-22> (5)
>    - *D* nbgrader/exporters/assignmentexporter.py
>    <https://github.com/jupyter/nbgrader/pull/154/files#diff-23> (68)
>    - *D* nbgrader/exporters/feedbackexporter.py
>    <https://github.com/jupyter/nbgrader/pull/154/files#diff-24> (18)
>    - *M* nbgrader/html/formgrade.py
>    <https://github.com/jupyter/nbgrader/pull/154/files#diff-25> (7)
>    - *D* nbgrader/preprocessors/findstudentid.py
>    <https://github.com/jupyter/nbgrader/pull/154/files#diff-26> (53)
>    - *M* nbgrader/preprocessors/saveautogrades.py
>    <https://github.com/jupyter/nbgrader/pull/154/files#diff-27> (18)
>    - *M* nbgrader/preprocessors/savecells.py
>    <https://github.com/jupyter/nbgrader/pull/154/files#diff-28> (18)
>    - *D* nbgrader/tests/test_findstudentid.py
>    <https://github.com/jupyter/nbgrader/pull/154/files#diff-29> (31)
>    - *M* nbgrader/tests/test_nbgrader_assign.py
>    <https://github.com/jupyter/nbgrader/pull/154/files#diff-30> (125)
>    - *M* tasks.py
>    <https://github.com/jupyter/nbgrader/pull/154/files#diff-31> (4)
>
> Patch Links:
>
>    - https://github.com/jupyter/nbgrader/pull/154.patch
>    - https://github.com/jupyter/nbgrader/pull/154.diff
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jupyter/nbgrader/pull/154>.
>



-- 
Brian E. Granger
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

jhamrick commented:
Ok @ellisonbg I think I am fairly happy with where this is at now. I still need to flesh out the docs, but as this is already a pretty massive PR I'm going to save that for a separate one. I did update the command line help for the all the apps, so there's at least some reference for what's going on.

I also updated the config stuff a bit, to make it so there only needs to be one config file, `nbgrader_config.py`, regardless of the subapp. I also added separate configurable classes that contain the common config options (both those inherited from IPython, and the nbgrader stuff like `assignment_id`) so that they only get listed once, rather than per-application. I think this makes it clearer what config options need to be changed to get the desired functionality.

jhamrick commented:
I am starting to migrate the code for my class to work with this PR. I am finding things here and there that still need to be addressed (and opening issues for them), but in general this is working *so* much better than nbgrader did before. Being able to (1) have a single config file for all the different steps and (2) have it know what to do just based on the directory structure is *huge*.

Carreau commented on nbgrader/apps/assignapp.py:
```diff
@@ -12,103 +13,152 @@
+    ),
+    'create': (
+        {'AssignApp': {'create_assignment': True}},
+        "Create the assignment at runtime if it does not exist."
```
Not sure what you mean by "at runtime"

jhamrick replied:
How about "Create an entry for the assignment in the database, if one does not already exist"?

Carreau commented on nbgrader/apps/assignapp.py:
```diff
@@ -12,103 +13,152 @@
+               students' solutions, or that contain students' written answers.
+               This is done so that we can warn students if they have changed
+               the tests, or if they have failed to provide a response to a
+               written answer.
```
I would flip the sentense around. Most user don't care if it computes checksum, which is an implementation detail. They want to know that changed cell will raise a warning.

Carreau commented on nbgrader/apps/autogradeapp.py:
```diff
@@ -35,76 +31,96 @@ class AutogradeApp(BaseNbConvertApp):
+    )
+
+    nbgrader_step_input = Unicode(
+        "submitted",
```
This name feel a lot like a `Bool`. I see what you mean, you wan it to match `/{submitted}/` pattern. I don't have a better name though.

jhamrick replied:
Yeah, I'm not a big fan of the name either. How about `nbgrader_input_step_name`?

Carreau commented on nbgrader/apps/baseapp.py:
```diff
@@ -52,26 +71,90 @@ class BaseNbGraderApp(BaseIPythonApplication):
+        return [BasicConfig]
+
+    def _config_file_name_default(self):
+        return u'nbgrader_config.py'
```
I have to check, but you "might" be able to drop the `.py` now that we have also json files.
Not sure.

jhamrick replied:
Ok, I'll check.

Carreau commented on nbgrader/apps/baseapp.py:
```diff
@@ -88,13 +171,99 @@ class BaseNbConvertApp(BaseNbGraderApp, NbConvertApp):
     flags = Dict(nbconvert_flags)
 
+    use_output_suffix = Bool(False)
+    postprocessor_class = DottedOrNone('')
```
Note (to self, also) with `DottedOrNone` you might need to import yourself, see if `Type` would work too and simplify the logic by auto-importing.

jhamrick replied:
Yeah, I think you do need to import it yourself, but not if you use `Type`. Though in this case, it doesn't matter, because there's nothing to import.

Carreau commented on nbgrader/apps/feedbackapp.py:
```diff
@@ -37,38 +26,74 @@ class FeedbackApp(BaseNbConvertApp):
 
     examples = Unicode(dedent(
```
Also, maybe dumb question, but what are the use of making this traitlets if it's not configurable ? I'm not sure where it is used...

jhamrick replied:
Not a dumb question, that's actually a good question. I'm not sure. The `examples` are something special having to do with `Application` I think (they are displayed when you do `--help`), but I don't know if they actually need to be traitlets.

Carreau replied:
Yep, you [are right])https://github.com/ipython/ipython/blob/master/IPython/config/application.py#L372-L383) if might not need to be a traitlet, nor be be detended IIUC.

jhamrick replied:
Ah, good point about it not needing to be dedented. I'll see what happens if I also don't make it a traitlet.

Carreau commented on nbgrader/apps/assignapp.py:
```diff
@@ -12,103 +13,152 @@
-            )
-        self.config.merge(self.extra_config)
+        extra_config = super(AssignApp, self).build_extra_config()
+        extra_config.NbGraderConfig.student_id = '.'
```
`dot` for a student id ? regular expression ?

jhamrick replied:
It's not a regex -- it's the actual value that gets put into the format string. It makes it so that directory isn't included, essentially. e.g. if the directory structure is `{nbgrader_step}/{student_id}/{assignment_id}/`, then for the assign step it will actually look for stuff in `source/./{assignment_id}/` and save stuff to `release/./{assignment_id}/`

Carreau replied:
Oh, I see. In my mind it was id like sql database ID... Sorry.

jhamrick replied:
Yeah, usually it is more like that, but in the case of `nbgrader assign`, there is no student. I don't entirely like that I've made it `'.'`, but I don't really see a better way of doing it.

Carreau commented:
I read through. I don't say i understand all the details, but that looks fine to me.

jhamrick commented:
Thanks for reviewing, @Carreau !
